Doc/library/typing.rst

```diff
@@ -968,520 +1075,479 @@ The module defines the following classes, functions and decorators:
+          def close(self): ...
 
-   * The first argument to ``Annotated`` must be a valid type
+      assert isinstance(open('/some/file'), Closable)
 
-   * Multiple type annotations are supported (``Annotated`` supports variadic
-     arguments)::
+   .. note::
 
-       Annotated[int, ValueRange(3, 10), ctype("char")]
+        :func:`runtime_checkable` will check only the presence of the required methods,
+        not their type signatures! For example, :class:`builtins.complex <complex>`
+        implements :func:`__float__`, therefore it passes an :func:`issubclass` check
+        against :class:`SupportsFloat`. However, the ``complex.__float__`` method
+        exists only to raise a :class:`TypeError` with a more informative message.
 
-   * ``Annotated`` must be called with at least two arguments (
-     ``Annotated[int]`` is not valid)
+   .. versionadded:: 3.8
 
-   * The order of the annotations is preserved and matters for equality
-     checks::
+.. decorator:: type_check_only
 
-       Annotated[int, ValueRange(3, 10), ctype("char")] != Annotated[
-           int, ctype("char"), ValueRange(3, 10)
-       ]
+   Decorator to mark a class or function to be unavailable at runtime.
 
-   * Nested ``Annotated`` types are flattened, with metadata ordered
-     starting with the innermost annotation::
+   This decorator is itself not available at runtime. It is mainly
+   intended to mark classes that are defined in type stub files if
+   an implementation returns an instance of a private class::
 
-       Annotated[Annotated[int, ValueRange(3, 10)], ctype("char")] == Annotated[
-           int, ValueRange(3, 10), ctype("char")
-       ]
+      @type_check_only
+      class Response:  # private or not available at runtime
+          code: int
+          def get_header(self, name: str) -> str: ...
 
-   * Duplicated annotations are not removed::
+      def fetch_response() -> Response: ...
 
-       Annotated[int, ValueRange(3, 10)] != Annotated[
-           int, ValueRange(3, 10), ValueRange(3, 10)
-       ]
+   Note that returning instances of private classes is not recommended.
+   It is usually preferable to make such classes public.
 
-   * ``Annotated`` can be used with nested and generic aliases::
 
-       T = TypeVar('T')
-       Vec = Annotated[List[Tuple[T, T]], MaxLen(10)]
-       V = Vec[int]
+Constant
+--------
 
-       V == Annotated[List[Tuple[int, int]], MaxLen(10)]
+.. data:: TYPE_CHECKING
+
+   A special constant that is assumed to be ``True`` by 3rd party static
+   type checkers. It is ``False`` at runtime. Usage::
+
+      if TYPE_CHECKING:
+          import expensive_mod
+
+      def fun(arg: 'expensive_mod.SomeType') -> None:
+          local_var: expensive_mod.AnotherType = other_fun()
+
+   Note that the first type annotation must be enclosed in quotes, making it a
+   "forward reference", to hide the ``expensive_mod`` reference from the
+   interpreter runtime.  Type annotations for local variables are not
```

Comment:
Unless `from __future__ import annotations` is used (PEP 563).

(Also, in 3.10 that will be the default. But we shouldn't change this until that's actually implemented -- that's bpo-38605, GH-20434.)

Reply:
Good point.

That reminds me of another matter: most of the generic types in `typing` will be redundant as their pre-existing corresponding types support `[]` in Python 3.9. [PEP 585—Type Hinting Generics In Standard Collections](https://www.python.org/dev/peps/pep-0585/) describes a "soft" deprecation process over 5 years after the 3.9 release.

I will think about how to add that information to the page.

Reply:
If you don't mind, I would prefer to land this PR first, then deal with the implications of PEP 585 and PEP 563 in a later PR (or two, since PEP 563 is still waiting for a PR). Especially the latter, since it won't apply to 3.9 (though a fair warning in 3.9 might still be useful).

Reply:
I added a note about PEP 563 in the entry about `typing.TYPE_CHECKING`.

Regarding the mass deprecation of PEP 585 I added a general note explaining the deprecation process to the top of section **Module Contents** and deprecation notes to each of the 37 affected types.